File: packages/python/unswallow/scan.py
from __future__ import annotations

import re
from typing import Dict, List, Tuple

from .types import ChannelSource

REASONING_FIELDS: Tuple[str, ...] = ("reasoning", "reasoning_content", "thinking", "thought")

THINK_TAG = re.compile(r"<\s*(\/?)\s*([a-zA-Z0-9_.:\-]*?\s*(?:think|response))[^>]*>", re.I)


class Region:
    __slots__ = ("channel", "source", "text")

    def __init__(self, channel: str, source: str, text: str) -> None:
        self.channel = channel
        self.source = source
        self.text = text


def split_think_blocks(text: str) -> List[Region]:
    out: List[Region] = []
    depth = 0
    last = 0
    block_start = 0
    for m in THINK_TAG.finditer(text):
        is_close = m.group(1) == "/"
        name = m.group(2).strip()
        is_response = re.search(r"response", name, re.I) is not None
        is_open_think = not is_close and not is_response
        if is_open_think:
            if depth == 0:
                if m.start() > last:
                    out.append(Region("content", "field", text[last : m.start()]))
                block_start = m.start()
            depth += 1
        else:
            if depth > 0:
                depth -= 1
                if depth == 0:
                    out.append(Region("thinking", "think-block", text[block_start : m.end()]))
                    last = m.end()
    if depth > 0:
        out.append(Region("thinking", "leak", text[block_start:]))
        last = len(text)
    if last < len(text):
        out.append(Region("content", "field", text[last:]))
    if not out and text:
        out.append(Region("content", "field", text))
    return out


def extract_regions(message: Dict, additional_fields=None) -> List[Region]:
    regions: List[Region] = []
    fields = list(REASONING_FIELDS) + list(additional_fields or [])
    for f in fields:
        v = message.get(f)
        if isinstance(v, str) and v:
            regions.append(Region(f, "field", v))
    content = message.get("content")
    if isinstance(content, str) and content:
        regions.extend(split_think_blocks(content))
    return regions

File: packages/unswallow/src/scan.ts
import type { ChannelSource, RawMessage } from './types';

export const REASONING_FIELDS: ChannelSource[] = [
  'reasoning',
  'reasoning_content',
  'thinking',
  'thought',
];

export interface Region {
  id: number;
  channel: ChannelSource;
  source: 'field' | 'think-block' | 'leak';
  text: string;
}

const THINK_TAG = /<\s*(\/?)\s*([a-zA-Z0-9_.:-]*?\s*(?:think|response))[^>]*>/gi;

export function splitThinkBlocks(text: string, startId = 0): Region[] {
  const out: Region[] = [];
  let nextId = startId;
  THINK_TAG.lastIndex = 0;
  let depth = 0;
  let last = 0;
  let blockStart = 0;
  let m: RegExpExecArray | null;
  while ((m = THINK_TAG.exec(text)) !== null) {
    const isClose = m[1] === '/';
    const name = m[2].trim();
    const isResponse = /response$/i.test(name) || /response/i.test(name);
    const isOpenThink = !isClose && !isResponse;
    if (isOpenThink) {
      if (depth === 0) {
        if (m.index > last) {
          out.push({ id: nextId++, channel: 'content', source: 'field', text: text.slice(last, m.index) });
        }
        blockStart = m.index;
      }
      depth++;
    } else {
      if (depth > 0) {
        depth--;
        if (depth === 0) {
          out.push({
            id: nextId++,
            channel: 'thinking',
            source: 'think-block',
            text: text.slice(blockStart, m.index + m[0].length),
          });
          last = m.index + m[0].length;
        }
      }
    }
  }
  if (depth > 0) {
    out.push({ id: nextId++, channel: 'thinking', source: 'leak', text: text.slice(blockStart) });
    last = text.length;
  }
  if (last < text.length) {
    out.push({ id: nextId++, channel: 'content', source: 'field', text: text.slice(last) });
  }
  if (out.length === 0 && text.length > 0) {
    out.push({ id: nextId++, channel: 'content', source: 'field', text });
  }
  return out;
}

export function extractRegions(message: RawMessage, additionalFields?: string[]): Region[] {
  const regions: Region[] = [];
  const fields = [...REASONING_FIELDS, ...(additionalFields ?? [])];
  for (const f of fields) {
    const v = (message as Record<string, unknown>)[f];
    if (typeof v === 'string' && v.length > 0) {
      regions.push({ id: regions.length, channel: f as ChannelSource, source: 'field', text: v });
    }
  }
  const content = typeof message.content === 'string' ? message.content : '';
  if (content.length > 0) {
    regions.push(...splitThinkBlocks(content, regions.length));
  }
  return regions;
}